### ISSUE TITLE
long running database connections.

### ISSUE THREAD

**Issue by mollybsmith-noaa** [NOAA-GSL/MATS#250]
---

Author Name: **randy.pierce** (@randytpierce)
Original Redmine Issue: 38453, https://vlab.ncep.noaa.gov/redmine/issues/38453
Original Date: 2017-09-28
Original Assignee: randy.pierce

---

From Bill....
Hi Randy,

I noticed the load on emb-prod-verif spiked recently, so I looked at processes. I saw a number of long-connected processes from MATS that almost certainly have nothing to do with the high load.

But a large number of idle connections can't be good for the db in general.  I suspect you have some jobs that fail to disconnect from the db after they're done, of if they go into some kind of error state.

I don't think this is urgent, but it would probably be a good thing to look into eventually.

-Bill
(see attached screen image)
	

---

- [image.png](https://vlab.ncep.noaa.gov/redmine/attachments/download/13523/image.png) (randy.pierce)

**Comment by mollybsmith-noaa** [comment]
---

Original Redmine Comment
Author Name: **molly.b.smith** (@mollybsmith-noaa)
Original Date: 2018-08-28T22:25:17Z

---

Issue 27872 seemed to think that this wasn't an issue. In the light of what we've discussed recently, would anyone like to comment?

**Comment by mollybsmith-noaa** [comment]
---

Original Redmine Comment
Author Name: **molly.b.smith** (@mollybsmith-noaa)
Original Date: 2018-09-07T00:24:39Z

---

This issue is from a year ago, but Bill pointed out on 9/5/18 that it's still occurring. I'm updating this thread with his email, and Randy's response.

From Bill:
I notice that MATS seems to be leaving a lot of open connections in the db. I don't know to what extent these slow down the database engine (at one point there was a modest load, and all db connections were in sleep), but if they accumulate enough, we could run out of database connections.

Is there anything relatively easy that might be done to avoid these long-sleeping processes?

-Bill

mysql> show processlist;
+--------+---------------+----------------------------------------+------------------+---------+-------+-------+------------------+
| Id     | User          | Host                                   | db               | Command | Time  | State | Info             |
+--------+---------------+----------------------------------------+------------------+---------+-------+-------+------------------+
| 191666 | readonly      | mats-dev.gsd.esrl.noaa.gov:45116       | ceiling2         | Sleep   | 12193 |       | NULL             |
| 191667 | readonly      | mats-dev.gsd.esrl.noaa.gov:45118       | ceiling_sums2    | Sleep   | 12204 |       | NULL             |
| 191668 | readonly      | mats-dev.gsd.esrl.noaa.gov:45120       | mats_common      | Sleep   | 12193 |       | NULL             |
| 199154 | loader2_madis | emb-prod-verif.gsd.esrl.noaa.gov:38986 | anom_corr2       | Sleep   | 13616 |       | NULL             |
| 223437 | readonly      | mats-dev.gsd.esrl.noaa.gov:45632       | ceiling_sums2    | Sleep   | 12193 |       | NULL             |
| 254357 | readonly      | mats-dev.gsd.esrl.noaa.gov:46056       | ruc_ua           | Sleep   |  8767 |       | NULL             |
| 254358 | readonly      | mats-dev.gsd.esrl.noaa.gov:46058       | mats_common      | Sleep   |  8767 |       | NULL             |
| 255789 | readonly      | mats-dev.gsd.esrl.noaa.gov:46182       | ruc_ua_sums2     | Sleep   |  8777 |       | NULL             |
| 255790 | readonly      | mats-dev.gsd.esrl.noaa.gov:46184       | ruc_ua_sums2     | Sleep   |  8777 |       | NULL             |
| 259200 | readonly      | mats.gsd.esrl.noaa.gov:49312           | anom_corr2       | Sleep   |  6703 |       | NULL             |
| 259201 | readonly      | mats.gsd.esrl.noaa.gov:49314           | anom_corr2       | Sleep   |  6043 |       | NULL             |
| 281096 | readonly      | mats-dev.gsd.esrl.noaa.gov:46462       | surfrad3         | Sleep   |  9145 |       | NULL             |
| 281097 | readonly      | mats-dev.gsd.esrl.noaa.gov:46464       | surfrad3         | Sleep   |  9145 |       | NULL             |
| 295245 | readonly      | mats-dev.gsd.esrl.noaa.gov:46518       | mats_common      | Sleep   |  8425 |       | NULL             |
| 295246 | readonly      | mats-dev.gsd.esrl.noaa.gov:46520       | vil              | Sleep   |  8425 |       | NULL             |
| 295247 | readonly      | mats-dev.gsd.esrl.noaa.gov:46524       | vil              | Sleep   |  8438 |       | NULL             |
| 295248 | readonly      | mats-dev.gsd.esrl.noaa.gov:46526       | vil              | Sleep   |  8434 |       | NULL             |
| 295250 | readonly      | mats-dev.gsd.esrl.noaa.gov:46530       | vil              | Sleep   |  8434 |       | NULL             |
| 297355 | readonly      | mats.gsd.esrl.noaa.gov:50122           | mats_common      | Sleep   |  6043 |       | NULL             |
| 298788 | readonly      | mats-dev.gsd.esrl.noaa.gov:46826       | mats_common      | Sleep   |  8072 |       | NULL             |
| 298789 | readonly      | mats-dev.gsd.esrl.noaa.gov:46828       | visibility       | Sleep   |  8072 |       | NULL             |
| 298796 | readonly      | mats-dev.gsd.esrl.noaa.gov:46830       | visibility_sums2 | Sleep   |  8072 |       | NULL             |
| 298933 | readonly      | mats-dev.gsd.esrl.noaa.gov:46832       | visibility_sums2 | Sleep   |  8082 |       | NULL             |
| 360843 | wcron0_user   | fe7.boulder.rdhpcs.noaa.gov:50616      | cref             | Sleep   |  2737 |       | NULL             |
| 369198 | readonly      | mats.gsd.esrl.noaa.gov:51494           | mats_common      | Sleep   |   801 |       | NULL             |
| 369200 | readonly      | mats.gsd.esrl.noaa.gov:51496           | precip2          | Sleep   |   775 |       | NULL             |
| 369201 | readonly      | mats.gsd.esrl.noaa.gov:51498           | precip2          | Sleep   |   775 |       | NULL             |
| 384262 | UA_retro      | tfe02.fairmont.rdhpcs.noaa.gov:60660   | ruc_ua_sums2     | Sleep   |    12 |       | NULL             |


From Randy:
So here is the issue.
Each Meteor app uses a database connection pool with a connection limit of 10. Actually it may use two pools one for data and one for metadata, the data one has 10 connections and the metadata one has 1. Usually these connections are well managed by the pool. 

One possibility could happen when the server (in this case nginx/passenger-phusion) is abruptly killed or restarted or something even nastier. I don't yet know of a way to catch a server termination event from within the app and then take a little time to clean up the connection pools before expiring - sort of a last gasp kind of thing. I have to look into this. The problem might be happening on mats-dev because the nightly build does do a restart of the web server. I wonder about though, because the longest "Time" on this list appears to be 3 hours.

Molly is probably correct that these connections are initiated by the automated tests. When they remain unused it seems that the underlying pool should close them after some default time. I don't know what that default time is or should be. 

In looking into this I did discover that the mysql client package that we are using is pretty old and that there is a newer version of an npm client available. Molly and I should update that package in the refactoring that we are planning in the near future. Perhaps it will then handle these persistent connections better.

If it doesn't, we may need to implement something that closes idle connections manually.

more investigation is required, I think.

randy


Again from Randy:
I believe that I have learned how to implement the database connections in each app in such a way that connection pools that are idle for more than a specified amount of time automatically get closed. This should largely alleviate this problem. I think it is a relatively small change.
randy

**Comment by mollybsmith-noaa** [comment]
---

Original Redmine Comment
Author Name: **molly.b.smith** (@mollybsmith-noaa)
Original Date: 2018-09-07T20:01:40Z

---

Follow-up from Randy:

I've been looking into this and here is what I've found.

Each MATS app generally opens two data connections for it's data pool for each server instance that is running plus a different pool to the mats_common metadata database. The number of connections is limited to 10 but I don't think that we can ever really get to 10 connections per app, it's just the default limit. I think we generally get two, sometimes three. It is interesting to note that once we finish the planned data model and move our app parameter generation to ingest time the connection count will probably go to one per app per server. We could reduce it to one per app per server now by using a single connection for the metadata and then destroying it when we finish loading the metadata and leaving the pool only for backend data requests, but it might not be worth the refactoring right now, especially since we can do that for each app as we migrate the metadata processing to the ingest side for the new data model. 

The pool connections are cached so that they don't have to be re-established for every request since establishing mysql connections is pretty expensive. I looked into the impact on mysql performance of having idle connections and it appears to be minimal (assuming that the server is not compiled for debug - which I hope ours is not) for up to hundreds of connections. This is what leaves me to think that we are better off leaving the pools to operate the way they currently are - i.e. keeping two, sometimes three idle connections per app per server. 

Our mysql server (emb-proddb.gsd.esrl.noaa.gov) is using the default wait_timeout of 8 hours. This means that idle connections will only live for eight hours at most anyway. We can see that because right now we only see around twenty two connections for the entire production server and 5 of those are to mats_common (metadata queries).

If we get a little extra time we can recode the 14 or so main.js files to stop using a pool for metadata on each app and that will leave us with fewer connections.
randy
